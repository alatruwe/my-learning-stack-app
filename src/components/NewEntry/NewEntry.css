form p {
  font-size: calc(1rem + 2vw);
}

.mood-date {
  display: flex;
  flex-direction: column;
  margin-bottom: 10%;
}

.new-entry-form-date {
  padding-bottom: 10%;
  font-size: calc(1rem + 2vw);
}

.mood-buttons {
  display: flex;
  justify-content: space-around;
  padding-top: 10%;
}

.active-mood {
  transform: scale(2);
  transition: all 0.5s ease-in-out;
  margin-right: 5%;
  padding: 0.2em 0.5em;
  background-color: transparent;
  border-radius: 10px;
  border: 1px solid rgba(255, 255, 255, 0.18);
}

.active-tech {
  transform: scale(1.2);
  transition: all 0.5s ease-in-out;
  margin-right: 5%;
  padding: 0.2em 0.5em;
}

.fa-meh {
  color: #ff8700;
}

.fa-smile {
  color: #0aff99;
}

.fa-frown {
  color: #e6403a;
}

.tech-buttons {
  margin-bottom: 10%;
}

.tech-button {
  text-decoration: none;
  margin: 5% 5%;
}

.profile-list {
  display: flex;
  flex-flow: row wrap;
  list-style-type: none;
}

.form-btn button {
  margin-bottom: 10%;
}

/* Desktop view */
@media screen and (min-width: 1024px) {
  .mood-date {
    margin: 0;
  }
  .mood-buttons {
    padding: 0;
  }
  .new-entry-form-date {
    padding-bottom: 0%;
    font-size: calc(1rem + 1vw);
  }
  .new-entry-text {
    font-size: calc(1rem + 1vw);
    padding-bottom: 2%;
  }
  .tech-buttons {
    padding-top: 2%;
    padding-bottom: 2%;
    margin: 0;
  }
  .tech-button {
    margin: 2% 2%;
  }
}
