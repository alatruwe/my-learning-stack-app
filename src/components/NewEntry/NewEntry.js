import React, { Component } from "react";
import EntriesApiService from "../../services/entries-api-services";
import ProfileApiService from "../../services/profile-api-services";
import ValidationError from "../ValidationError/ValidationError.js";
import Tech from "../Tech/Tech.js";
import "./NewEntry.css";

export default class NewEntry extends Component {
  constructor(props) {
    super(props);
    var today = new Date(),
      date =
        today.getMonth() +
        1 +
        "-" +
        today.getDate() +
        "-" +
        today.getFullYear();
    this.state = {
      techs: [],
      date: date,
      mood: "",
      tech: "",
      learningNotes: "",
      strugglingNotes: "",
      error: null,
    };
  }
  static defaultProps = {
    history: {
      push: () => {},
    },
  };

  // form state updates
  updateLearningNotes = (notes) => {
    this.setState({ learningNotes: notes });
  };
  updateStrugglingNotes = (notes) => {
    this.setState({ strugglingNotes: notes });
  };

  updateTech = (name) => {
    this.setState({ tech: name });
  };

  updateMood = (id) => {
    this.setState({ mood: id });
  };

  // form vaidation
  validateEntry = () => {
    const mood = this.state.mood;
    const tech = this.state.tech;

    if (mood === "" || tech === "") {
      return "Please enter at least your mood and your focus";
    }
  };

  // get profile to display correct tech
  componentDidMount() {
    ProfileApiService.getProfile().then((res) => {
      this.setState({ techs: res });
    });
  }

  // handle submit
  handleSubmit = (e) => {
    e.preventDefault();
    // get media info
    const tech = this.state.tech;
    const current_mood = this.state.mood;
    const learning_notes = this.state.learningNotes;
    const struggling_notes = this.state.strugglingNotes;
    let entry = {
      tech_id: tech,
      current_mood: current_mood,
      learning_notes: learning_notes,
      struggling_notes: struggling_notes,
    };
    // API POST request
    EntriesApiService.postEntry(entry)
      // reset form fields
      .then(() => {
        e.target.learning_notes.value = "";
        e.target.struggling_notes.value = "";
      })
      // change state
      .then(() => {
        this.setState({
          mood: "",
          tech: "",
          learningNotes: "",
          strugglingNotes: "",
          error: null,
        });
        this.props.history.push(`/dashboard`);
      })
      .catch((res) => {
        this.setState({ error: res.error });
      });
  };

  render() {
    const techs = this.state.techs;
    const entryError = this.validateEntry();
    return (
      <div className="wrapper">
        <section className="card">
          <h2 className="card-title">New Entry</h2>
          <div className="form">
            <form className="form-details" onSubmit={this.handleSubmit}>
              <div className="mood-date">
                <div className="new-entry-form-date">{this.state.date}</div>
                <div>
                  <p className="new-entry-text">I felt: </p>
                  <div className="mood-buttons">
                    <button
                      onClick={(e) => this.updateMood(e.currentTarget.id)}
                      type="button"
                      id="smile"
                      className={
                        this.state.mood === "smile" ? "active-mood" : "btn"
                      }
                    >
                      <i className="far fa-smile"></i>
                    </button>
                    <button
                      onClick={(e) => this.updateMood(e.currentTarget.id)}
                      type="button"
                      id="meh"
                      className={
                        this.state.mood === "meh" ? "active-mood" : "btn"
                      }
                    >
                      <i className="far fa-meh"></i>
                    </button>
                    <button
                      onClick={(e) => this.updateMood(e.currentTarget.id)}
                      type="button"
                      id="frown"
                      className={
                        this.state.mood === "frown" ? "active-mood" : "btn"
                      }
                    >
                      <i className="far fa-frown"></i>
                    </button>
                  </div>
                </div>
              </div>
              <div className="tech-buttons">
                <p className="new-entry-text">I focused on:</p>
                <ul className="profile-list">
                  {techs.map((tech) => (
                    <li
                      key={tech.id}
                      className={
                        this.state.tech === tech.name
                          ? "active-tech"
                          : "tech-button"
                      }
                    >
                      <Tech
                        id={tech.name}
                        tech={tech.name}
                        updateTech={this.updateTech}
                      />
                    </li>
                  ))}
                </ul>
              </div>
              <div className="notes">
                <p className="new-entry-text">Notes:</p>
                <textarea
                  id="learning_notes"
                  name="learning_notes"
                  placeholder="I learned..."
                  className="rounded-input"
                  rows="5"
                  value={this.state.learningNotes}
                  onChange={(e) => this.updateLearningNotes(e.target.value)}
                />
                <textarea
                  id="struggling_notes"
                  name="struggling_notes"
                  placeholder="I struggled with..."
                  className="rounded-input"
                  rows="5"
                  value={this.state.strugglingNotes}
                  onChange={(e) => this.updateStrugglingNotes(e.target.value)}
                />
              </div>

              <div className="form-btn">
                <button
                  type="submit"
                  className="btn"
                  disabled={this.validateEntry()}
                >
                  Add entry
                </button>
                {this.validateEntry() && (
                  <ValidationError message={entryError} />
                )}
              </div>
            </form>
          </div>
        </section>
      </div>
    );
  }
}
