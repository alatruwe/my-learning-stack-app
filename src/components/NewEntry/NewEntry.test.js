import React from "react";
import ReactDOM from "react-dom";
import NewEntry from "./NewEntry";
import renderer from "react-test-renderer";

describe(`NewEntry component`, () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<NewEntry />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer.create(<NewEntry />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
