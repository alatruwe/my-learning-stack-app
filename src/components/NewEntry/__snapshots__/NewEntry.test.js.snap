// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewEntry component renders the UI as expected 1`] = `
<section
  className="wrapper"
>
  <h1
    className="form-title"
  >
    <span>
      New Entry
    </span>
  </h1>
  <div
    className="form"
  >
    <form
      className="form-details"
      onSubmit={[Function]}
    >
      <div>
        <div>
          <button
            className="btn"
            id="smile"
            onClick={[Function]}
            type="button"
          >
            <i
              className="far fa-smile"
            />
          </button>
          <button
            className="btn"
            id="meh"
            onClick={[Function]}
            type="button"
          >
            <i
              className="far fa-meh"
            />
          </button>
          <button
            className="btn"
            id="frown"
            onClick={[Function]}
            type="button"
          >
            <i
              className="far fa-frown"
            />
          </button>
        </div>
        <div>
          3-24-2021
        </div>
      </div>
      <div>
        <button
          className="btn"
          id="react"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fab fa-react"
          />
        </button>
        <button
          className="btn"
          id="javascript"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fab fa-js"
          />
        </button>
        <button
          className="btn"
          id="node.js"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fab fa-node-js"
          />
        </button>
      </div>
      <div>
        <textarea
          className="rounded-input"
          id="learning-notes"
          name="learning-notes"
          onChange={[Function]}
          placeholder="I learned..."
          rows="5"
          value=""
        />
        <textarea
          className="rounded-input"
          id="struggling-notes"
          name="struggling-notes"
          onChange={[Function]}
          placeholder="I struggled with..."
          rows="5"
          value=""
        />
      </div>
      <div
        className="form-btn"
      >
        <button
          className="btn"
          disabled="Please enter your mood and your tech"
          type="submit"
        >
          Add entry
        </button>
        <div
          className="error"
        >
          Please enter your mood and your tech
        </div>
      </div>
    </form>
  </div>
</section>
`;
